### PR TITLE
add support for new GitHub authorizations API using fingerprint

### LINKED ISSUE #182: OAuth error when using same GitHub account on multiple computers
I have multiple workstations with the same configuration using the same GitHub account to authenticate. Recently, after running Boxen on one computer, the remaining computers are unable to use the token. The only way to run Boxen again is to delete the account's personal application token on GitHub.

```
--> Preparing to auto-update...
--> Complete! Nothing new from upstream.
--> Oh, looks like you've provided your login as environmental variable...

GitHub password: *****************

--> Something went terribly wrong.
    I was able to get your OAuth token, but was unable to use it.
```

### PR DESCRIPTION
**Pull request by ptoomey3**
This PR attempts to fix the issues discussed in https://github.com/boxen/boxen/issues/182. The general approach is:

* Detect if there is any existing authorization that has the same description and fingerprint.
* If there is a match we delete the existing authorization, as we can't retrieve its plaintext token value.
* Create a new authorization using the `SHA256` of the user's machine as the fingerprint and the user's hostname as part of the description (to help users manage tokens in the GitHub UI).

This PR currently takes no steps to delete existing tokens that do not yet have a fingerprint (i.e. all existing tokens in use). Right now a user may have the same token on multiple installs of boxen. So, if we add logic to delete the single existing token it will invalidate it on all the user's other installs. This probably isn't a big deal, as it should simply ask them to auth again. But. I thought I'd double check that assumption with you all before adding the deletion logic. 

fixes #182

/cc @mastahyeti @gregose @jasonrudolph @pengwynn @mikemcquaid

### CODE REVIEW

**Comment by mastahyeti** [comment]
LGTM. Thanks for getting to this.

**Comment by ptoomey3** [comment]
Just realized I need to add support for OTP for the authorization delete operation. Also, I'm leaning toward deleting any existing Boxen authorizations without a fingerprint, as the benefits seem to outweigh the downsides for those that only have a single install (probably most users). I'll double check to see what the user experience is like for users that haven't upgraded Boxen on at least one of their installs.

File: lib/boxen/preflight/creds.rb [hunk 1]
```diff
@@ -105,4 +110,24 @@ def fetch_from_env(thing)
     warn "Oh, looks like you've provided your #{thing} as environmental variable..."
     found
   end
+
+  def fingeprint
```
Comment (mikemcquaid) [inline lib/boxen/preflight/creds.rb]: Should this be called `fingerprint` instead?

**Comment by mikemcquaid** [comment]
@ptoomey3 Nice work here, thanks. Might want @jbarnette to chime in if he has any thoughts. The approach seems sane to me. Given that when this gets pushed out it won't affect all Boxen instances I think we can considering deleting existing tokens if that makes more security sense.

File: lib/boxen/preflight/creds.rb (hunk 1 above)
Comment (ptoomey3) [inline lib/boxen/preflight/creds.rb]: Yup :smile: - had that typo fixed up on my local branch but haven't pushed it up yet.

**Comment by ptoomey3** [comment]
> I think we can considering deleting existing tokens if that makes more security sense.

It is a small security benefit and it will, in general, be less confusing for users. If we don't delete the old legacy token then it will stick around until a user happens to go through their PATs and see that there is an old Boxen authorization that hasn't been used for some time.

File: lib/boxen/preflight/creds.rb [hunk 2]
```diff
@@ -105,4 +122,28 @@ def fetch_from_env(thing)
...
+      # Computer."
+      serial_number_match_data = IO.popen(
+        ["ioreg", "-c", "IOPlatformExpertDevice", "-d", "2"]
+      ).read.match(/"IOPlatformSerialNumber" = "([[:alnum:]]+)"/)
+      if serial_number_match_data
+        # The fingerprint must be unique across all personal access tokens for a
+        # given user. We prefix the serial number with the application name to
+        # differeniate between any other personal access token that uses the
+        # Mac serial number for the fingerprint.
+        Digest::SHA256.hexdigest("Boxen: #{serial_number_match_data[1]}")
```
Comment (ptoomey3) [inline lib/boxen/preflight/creds.rb]: The comment above explains the reasoning here. For normal OAuth apps we never have to worry about this, since the `fingerprint` must only be unique for a given client_id. But, with PATs there is no unique/real client_id, so `fingerprint` must be unique across all PATs.

File: lib/boxen/preflight/creds.rb [hunk 3]
```diff
@@ -68,12 +70,27 @@ def run
...
-        :note => "Boxen",
-        :scopes => %w(repo user),
-        :headers => headers
-    end
+    # Boxen now supports the updated GitHub Authorizations API by using a unique
+    # `fingerprint` for each Boxen installation for a user. We delete any older
+    # authorization that does not make use of `fingerprint` so that the "legacy"
+    # authorization doesn't persist in the user's list of personal access
+    # tokens.
+    legacy_auth = tokens.detect { |a| a.note == "Boxen" && a.fingerprint == nil }
```
Comment (ptoomey3) [inline lib/boxen/preflight/creds.rb]: This will take care of the base case for most users. But, there are edge cases we can't solve here. For example:

* User updates to this new version of boxen on Mac1 and logs in. 
* Legacy token they already had is deleted and a new token is created specifically for Mac1.
* Mac2 logs in using an older version of Boxen, and ends up creating yet a new "legacy token" that has no fingerprint.
* Mac3 logs in using an older version of Boxen and can't get the plaintext legacy token that Mac2 ended up creating. 
* Mac3 is stuck unless/until they either delete the other legacy token or upgrade their version of Boxen.

So, while we can't solve all problems with this change, it will be ideal for most users and other users will be no worse off than if we didn't try to clear the legacy token for them.

**Comment by ptoomey3** [comment]
Ok, I am no passing the OTP header consistently, deleting the legacy token, and I added a number of tests. The tests are not awesome, as so much of the API bits have to be mocked/stubbed. But, let me know if there is anything you would like to see added/changed.

**Comment by ptoomey3** [comment]
Also, it looks like highline has [officially abandoned support](https://github.com/JEG2/highline/issues/129#issuecomment-75893373) for Ruby 1.8.X., which explains why CI is failing for 1.8.7.

**Comment by seanknox** [comment]
@ptoomey3 :+1: looks good (thanks for the tests!) and your approach sounds good to me. 

Re: those edge cases you describe, I agree that this PR is certainly better for most people and that I think the best solution is to note in the FAQ, release notes, README, etc., the need to upgrade Boxen due to GitHub API changes.

**Comment by mikemcquaid** [comment]
@seanknox I think the release notes are enough.
@ptoomey3 Could you suggest a bit of release notes blurb to add when this is merged? Also, any chance of using an older version of highline as we're still trying to support 1.8 (for OS X 10.8).

**Comment by ptoomey3** [comment]
Ok, I adjusted the Gemfile to stick to highline < 1.7.0. And, here is a suggested blurb:

* Adds support for the [updated GitHub Authorizations API](https://developer.github.com/changes/2014-12-08-removing-authorizations-token/). This was a breaking API change and all Boxen users should update.

**Comment by ptoomey3** [comment]
I just tested out the changes hitting the actual API by doing instantiating as follows:

``` ruby
preflight = Boxen::Preflight::Creds.new @config
preflight.run
```

I tested the above with various scenarios, including:

* Existing legacy authorization - This deleted the existing legacy authorization and created a new  fingerprinted one named after my hostname.
* Existing fingerprinted authorization - This deleted the existing fingerprinted authorization and created a new one with the same name.
* After the legacy authorization was deleted I used an older copy of Boxen - This created a legacy token named "Boxen".
* After rerunning an older version of Boxen I ran the new one again - This both deleted the legacy authorization and replaced the existing fingerprinted authorization with a new one.

So, in short, I think this is all working the way we want. If someone else with merge rights wants to test before merging feel free. Otherwise, I *think* this should be good to go.

**Comment by mikemcquaid** [comment]
Great, thanks @ptoomey3. Will test and likely merge on Monday.

File: lib/boxen/preflight/creds.rb [hunk 4]
```diff
@@ -105,4 +122,28 @@ def fetch_from_env(thing)
     warn "Oh, looks like you've provided your #{thing} as environmental variable..."
     found
   end
+
+  def fingerprint
+    @fingerprint ||= begin
+      # See Apple technical note TN1103, "Uniquely Identifying a Macintosh
+      # Computer."
+      serial_number_match_data = IO.popen(
+        ["ioreg", "-c", "IOPlatformExpertDevice", "-d", "2"]
```
Comment (ptoomey3) [inline lib/boxen/preflight/creds.rb]: @mikemcquaid - I saw your comment about supporting OS X 10.8. I have no clue whether this command is supported in 10.8, but I believe ioreg itself has been in OS X for quite some time.

**Comment by mikemcquaid** [comment]
Testing looks good. Thanks @ptoomey3!
